[triangle] Add new port

n4z4m3 opened the pull request:
Add the libigl-triangle port as per Issue #12564
Fixes #12564

Commit:
New Port : libigl-triangle

Commit:
Update ci.baseline.txt for libigl-triangle port

NancyLi1013 commented on ports/libigl-triangle/portfile.cmake:
```diff
@@ -0,0 +1,28 @@
+    PREFER_NINJA
+)
+
+set(VCPKG_POLICY_EMPTY_INCLUDE_FOLDER enabled)
```
Why do you need to add this?

nazame replied:
You are right, this is not needed. I needed it before figuring out how to install the triangle.h file properly.

NancyLi1013 commented on ports/libigl-triangle/portfile.cmake:
```diff
@@ -0,0 +1,28 @@
+
+vcpkg_copy_pdbs()
+
+file(INSTALL ${CMAKE_CURRENT_LIST_DIR}/copyright DESTINATION ${CURRENT_PACKAGES_DIR}/share/libigl-triangle)
```
```suggestion
file(INSTALL ${SOURCE_PATH}/README DESTINATION ${CURRENT_PACKAGES_DIR}/share/${PORT} RENAME copyright)
```
If using `README` as copyright, we should update like this.

NancyLi1013 commented on scripts/ci.baseline.txt:
```diff
@@ -729,6 +729,8 @@ libidn2:x64-osx=fail
 libigl:arm-uwp=fail
 libigl:x64-uwp=fail
+libigl-triangle:arm-uwp=fail
+libigl-triangle:x64-uwp=fail
```
Does the port not support `arm `and `uwp`?
If so, we should add `Supports: !(arm|uwp) `to CONTROL file and also need to add `vcpkg_fail_port_install(ON_ARCH "arm" ON_TARGET "uwp")` to the top of portfile.cmake.

nazame replied:
It fails uwp so I added your suggestions. However, are your suggestions correct? Because I believe it supports arm, but not arm-uwp or x64-uwp so I think it is actually just uwp that is not supported.

NancyLi1013 replied:
Sorry, it's my mistake. It should be uwp that is not supported.

nazame commented on ports/libigl-triangle/CONTROL:
```diff
@@ -0,0 +1,4 @@
+Source: libigl-triangle
+Version: 1.6
+Port-Version: 1
```
I see. Thanks. Is it harmful to include it?

n4z4m3 committed:
Update ports/libigl-triangle/CONTROL

Unnecessary Port-Version

Co-authored-by: NancyLi1013 <46708020+NancyLi1013@users.noreply.github.com>

n4z4m3 committed:
Update ports/libigl-triangle/portfile.cmake

Unnecessary inclusion of vcpkg_common_functions

Co-authored-by: NancyLi1013 <46708020+NancyLi1013@users.noreply.github.com>

n4z4m3 committed:
Update ports/libigl-triangle/portfile.cmake

Use README from source as the copyright file

Co-authored-by: NancyLi1013 <46708020+NancyLi1013@users.noreply.github.com>

Commit:
Updates to libigl-translate port

Commit:
Update libigl-triangle port

cenit commented:
Can you please use official sources and not unofficial repository?
http://www.netlib.org/voronoi/triangle.zip

cenit commented:
Also, you install the header both in the CMakeLists and in the portfile. One should be enough :)

NancyLi1013 commented on ports/libigl-triangle/CONTROL:
```diff
@@ -0,0 +1,5 @@
+Version: 1.6
+Homepage: https://github.com/libigl/triangle
+Description: A Two-Dimensional Quality Mesh Generator and Delaunay Triangulator. Port from Jonathan Shewchuk's code which adds 64-bit support and the ability to build as library.
+Supports: !(arm|uwp)
```
```suggestion
Supports: !uwp
```

NancyLi1013 commented on ports/libigl-triangle/portfile.cmake:
```diff
@@ -0,0 +1,24 @@
+vcpkg_fail_port_install(ON_ARCH "arm" ON_TARGET "uwp")
```
```suggestion
vcpkg_fail_port_install(ON_TARGET "uwp")
```

NancyLi1013 commented on ports/libigl-triangle/portfile.cmake:
```diff
@@ -0,0 +1,24 @@
+
+vcpkg_copy_pdbs()
+
+file(INSTALL "${SOURCE_PATH}/triangle.h" DESTINATION ${CURRENT_PACKAGES_DIR}/include)
```
```suggestion
```
As @cenit said, the header has been installed in CMakeLists.txt. So there is no need to install it again here.

nazame replied:
Actually, I cannot do this. When I do I get the following error "The folder /include is empty or not present. This indicates the library was not correctly installed." For some reason, the CMakeLists.txt is not doing the copy to the include folder in packages. I don't know why. If you can help figure that out I will remove this line in the portfile.

cenit replied:
let's fix the CMakeLists.txt and not install it in the portfile :)

I will look into it later

nazame replied:
I have made a fix for this. However, it requires me to add file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include") to the portfile

cenit commented:
Final note: can you please also provide CMake targets, to link to triangle (prefixed with unofficial is ok if you prefer)

NancyLi1013 commented:
> Can you please use official sources and not unofficial repository?
> http://www.netlib.org/voronoi/triangle.zip

It seems the repository is not unofficial repository. I noticed that `libigl` also used https://github.com/libigl/libigl.
But I'm also not sure about this.

@BillyONeal 
Could you please help review this point? Also for the port name, I don't know if `libigl-triangle `is suitable. Or we might rename it as `triangle`.

cenit commented:
> It seems the repository is not unofficial repository

It is. I am also a user of triangle :)
Please look at http://www.cs.cmu.edu/~quake/triangle.html for the official homepage and official download link

cenit commented:
It is really the same sources, nothing looks like might warrant even a libigl-triangle duplicate here

n4z4m3 committed:
Update ports/libigl-triangle/CONTROL

Co-authored-by: NancyLi1013 <46708020+NancyLi1013@users.noreply.github.com>

n4z4m3 committed:
Update ports/libigl-triangle/portfile.cmake

Co-authored-by: NancyLi1013 <46708020+NancyLi1013@users.noreply.github.com>

nazame commented:
> Can you please use official sources and not unofficial repository?
> http://www.netlib.org/voronoi/triangle.zip

I would like to. The reason I did not is because this is actually not a true mirror of the original sources. It has modifications to the source code that allow it to build as a library and also to support 64-bit architecture. However, I think I should be able to add these changes using patches. I will have to learn how to do the patching. Hopefully the documentation at https://vcpkg.readthedocs.io/en/latest/examples/patching will help.

Also, this should allow me to change the port name to just triangle.

cenit commented:
>  I think I should be able to add these changes using patches. I will have to learn how to do the patching. Hopefully the documentation at https://vcpkg.readthedocs.io/en/latest/examples/patching will help.

yes please.
You will receive support in case of necessity

nazame commented:
> Final note: can you please also provide CMake targets, to link to triangle (prefixed with unofficial is ok if you prefer)

I would like to, but I am pretty new to both creating vcpkg ports (which is probably pretty obvious by now) and CMake. So I don't really know how to do this. But I will try to figure it out. If you or somebody else can help that would be great. Even just pointing to a good example that I can copy from would be helpful.

Commit:
Make the header copy properly and add CMake targets

cenit commented:
Yeah right direction!

cenit commented on ports/libigl-triangle/portfile.cmake:
```diff
@@ -0,0 +1,26 @@
+vcpkg_copy_pdbs()
+
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include")
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/share")
```
Instead of removing debug/share, you should use the vcpkg_fixup_cmake_targets() function

nazame replied:
Nice. I did not know about this. I did see this function used before, but could not figure out what it did.

Commit:
Utilize original sources for triangle library and rename to triangle

nazame commented:
I believe I have implemented all suggested changes.

I really appreciate all the suggestions. You are all awesome.

Note that there is one more change request which I believe is not going to be made. How do I close that? Or is that something that the person who made the request must do?

cenit commented:
Would you mind also create the triangle.exe target and move it to tools? That would make this a fully complete port (and in return you would also learn how to deal with executable tools in vcpkg ;) )

Commit:
Build triangle executable and add it to tools

nazame commented:
> Would you mind creating the triangle.exe target and move it to tools? That would make this a fully complete port (and in return you would also learn how to deal with executable tools in vcpkg ;) )

Yes. This has been done now.

NancyLi1013 commented:
The regressions on x64-linux:

```
4/5] : && /usr/bin/cc -fPIC -g  -rdynamic CMakeFiles/triangle.dir/triangle.c.o  -o triangle   && :
FAILED: triangle 
: && /usr/bin/cc -fPIC -g  -rdynamic CMakeFiles/triangle.dir/triangle.c.o  -o triangle   && :
CMakeFiles/triangle.dir/triangle.c.o: In function `parsecommandline':
/mnt/vcpkg-ci/buildtrees/triangle/src/triangle-84d2cb038f.clean/triangle.c:3527: undefined reference to `cos'
/mnt/vcpkg-ci/buildtrees/triangle/src/triangle-84d2cb038f.clean/triangle.c:3531: undefined reference to `sqrt'
CMakeFiles/triangle.dir/triangle.c.o: In function `circletop':
/mnt/vcpkg-ci/buildtrees/triangle/src/triangle-84d2cb038f.clean/triangle.c:10529: undefined reference to `sqrt'
CMakeFiles/triangle.dir/triangle.c.o: In function `splitencsegs':
/mnt/vcpkg-ci/buildtrees/triangle/src/triangle-84d2cb038f.clean/triangle.c:13362: undefined reference to `sqrt'
CMakeFiles/triangle.dir/triangle.c.o: In function `quality_statistics':
/mnt/vcpkg-ci/buildtrees/triangle/src/triangle-84d2cb038f.clean/triangle.c:15392: undefined reference to `cos'
/mnt/vcpkg-ci/buildtrees/triangle/src/triangle-84d2cb038f.clean/triangle.c:15501: undefined reference to `sqrt'
/mnt/vcpkg-ci/buildtrees/triangle/src/triangle-84d2cb038f.clean/triangle.c:15502: undefined reference to `sqrt'
/mnt/vcpkg-ci/buildtrees/triangle/src/triangle-84d2cb038f.clean/triangle.c:15503: undefined reference to `sqrt'
/mnt/vcpkg-ci/buildtrees/triangle/src/triangle-84d2cb038f.clean/triangle.c:15504: undefined reference to `sqrt'
/mnt/vcpkg-ci/buildtrees/triangle/src/triangle-84d2cb038f.clean/triangle.c:15510: undefined reference to `sqrt'
/mnt/vcpkg-ci/buildtrees/triangle/src/triangle-84d2cb038f.clean/triangle.c:15510: undefined reference to `acos'
/mnt/vcpkg-ci/buildtrees/triangle/src/triangle-84d2cb038f.clean/triangle.c:15516: undefined reference to `sqrt'
/mnt/vcpkg-ci/buildtrees/triangle/src/triangle-84d2cb038f.clean/triangle.c:15516: undefined reference to `acos'
/mnt/vcpkg-ci/buildtrees/triangle/src/triangle-84d2cb038f.clean/triangle.c:15518: undefined reference to `sqrt'
/mnt/vcpkg-ci/buildtrees/triangle/src/triangle-84d2cb038f.clean/triangle.c:15518: undefined reference to `acos'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```
Could you please look into this?

Commit:
Fix Linux build which requires linking to m library for math functions

nazame commented:
> The regressions on x64-linux:
> 
> Could you please look into this?

I believe I have fixed this. However, the checks are not re-running. Is that expected?

cenit commented:
Try rebasing on master to re-trigger CI checks

In any case, thanks for applying all suggestions. Port looks very fine for me :)

nazame commented:
> Try rebasing on master to re-trigger CI checks
> 
> In any case, thanks for applying all suggestions. Port looks very fine for me :)

Is this rebasing something I need to do? Because I have never had to do a rebase before and I am worried about messing something up.

Commit:
Merge branch 'master' of https://github.com/Microsoft/vcpkg into libigl-triangle

NancyLi1013 commented:
I have merged master to re-trigger CI checks. Let's see the new check results.

NancyLi1013 commented on scripts/ci.baseline.txt:
```diff
@@ -713,6 +713,8 @@ libidn2:x64-osx=fail
 libigl:arm64-windows=fail
 libigl:arm-uwp=fail
 libigl:x64-uwp=fail
+libigl-triangle:arm-uwp=fail
+libigl-triangle:x64-uwp=fail
```
```suggestion
```
These codes are not needed any more. Since you have added `Supports:!uwp ` to CONTROL file.

n4z4m3 committed:
Update scripts/ci.baseline.txt

Co-authored-by: NancyLi1013 <46708020+NancyLi1013@users.noreply.github.com>

ras0219-msft commented on ports/triangle/enable_64bit_architecture.patch:
```diff
@@ -0,0 +1,346 @@
+ 
++
++/* Define the size large enough to store and operate on a pointer.           */
++#define INT_PTR unsigned long long
```
This is suboptimal on 32-bit platforms -- can we instead simply `#include <stdint.h>` and use `intptr_t`?

nazame replied:
You are absolutely right. I have modified the patch file to make this change.

ras0219-msft reviewed (commented):
Thanks for the PR!

Other than my one comment, LGTM.

Commit:
Instead of defining INT_PTR, use uintptr_t from stdint.h
Also, define FLOAT and VOID in the header so the user of the library does not have to define them

Commit:
Make sure the port works and can be used

ras0219-msft commented on ports/triangle/enable_64bit_architecture.patch:
```diff
@@ -0,0 +1,854 @@
+-/*   compiler is smarter, feel free to replace the "int" with "void".        */
+-/*   Not that it matters.                                                    */
+-
+-#define VOID int
```
Is it necessary to remove this `#define` instead of changing it to `#define VOID void`? It significantly increases the size of this patch file.

nazame replied:
Changed this

ras0219-msft commented on ports/triangle/enable_64bit_architecture.patch:
```diff
@@ -0,0 +1,854 @@
+ /*                                                                           */
+ /*****************************************************************************/
+ 
++#ifdef SINGLE
```
Because we are only ever compiling the library in `double` mode, reacting to `SINGLE` in the consumer's code is just a footgun waiting to fire.

It would be better to replace `REAL` in this file with `double` directly.

nazame replied:
Done

Commit:
Remove the use of the SINGLE define for switching the REAL define to be float or double. Also make the define for VOID be void rather than replacing all VOID with void in order to reduce the patch size.

ras0219-msft commented:
Thanks for the PR!
